src/ints/int10_modes.cpp

```diff
@@ -1760,7 +1902,172 @@ uint32_t VideoModeMemSize(uint16_t mode) {
 	return static_cast<uint32_t>(mem_bytes);
 }
 
-void INT10_SetupPalette() {
-	init_all_palettes();
+static cga_colors_t handle_cga_colors_prefs_tandy(const std::vector<std::string> cga_colors_prefs)
+{
+	constexpr auto default_brown_level = 0.5f;
+
+	auto brown_level = default_brown_level;
+
+	if (cga_colors_prefs.size() > 1) {
+		auto brown_level_pref = cga_colors_prefs[1];
+
+		if (float p; sscanf(brown_level_pref.c_str(), "%f", &p)) {
+			auto cga_colors = cga_colors_default;
+
+			brown_level = clamp(p / 100, 0.0f, 1.0f);
+
+			constexpr float max_green = 0x2a;
+			cga_colors[6].green = static_cast<uint8_t>(max_green *
+			                                           brown_level);
+			return cga_colors;
+		} else {
+			LOG_WARNING("INT10: Invalid brown level value for 'tandy' CGA colors: '%s', "
+			            "using default brown level of %3.2f",
+			            brown_level_pref.c_str(),
+			            default_brown_level * 100);
+		}
+	}
+	return cga_colors_ibm5153;
+}
+
+static cga_colors_t handle_cga_colors_prefs_ibm5153(const std::vector<std::string> cga_colors_prefs)
+{
+	constexpr auto default_contrast = 1.0f;
+
+	auto contrast = default_contrast;
+
+	if (cga_colors_prefs.size() > 1) {
+		auto contrast_pref = cga_colors_prefs[1];
+
+		if (float p; sscanf(contrast_pref.c_str(), "%f", &p)) {
+			contrast = clamp(p / 100, 0.0f, 1.0f);
+
+			auto cga_colors = cga_colors_ibm5153;
+			for (size_t i = 0; i < cga_colors.size() / 2; ++i) {
+				// The contrast control effectively dims the
+				// first 8 non-bright colours only
+				auto c = cga_colors[i];
+				auto r = static_cast<float>(c.red) * contrast;
+				auto g = static_cast<float>(c.green) * contrast;
+				auto b = static_cast<float>(c.blue) * contrast;
+
+				cga_colors[i] = {static_cast<uint8_t>(r),
+				                 static_cast<uint8_t>(g),
+				                 static_cast<uint8_t>(b)};
+			}
+			return cga_colors;
+		} else {
+			LOG_WARNING("INT10: Invalid contrast value for 'ibm5153' CGA colors: '%s', "
+			            "using default contrast of %3.2f",
+			            contrast_pref.c_str(),
+			            default_contrast * 100);
+		}
+	}
+	return cga_colors_ibm5153;
```

Comment:
Suggest peeling the nested conditions off into stand-alone sentinels:

```c++

// Use has no preference, give them the default
if (cga_colors_prefs.size() <= 1)
    return cga_colors_ibm5153;

// Use provided a contrast level, if it's invalid give them the default
float p = 0.0f;
if (!sscanf(contrast_pref.c_str(), "%f", &p)) {
   LOG_WARNING ...
   return cga_colors_ibm5153;
}


// Sunny day scenario.. sentinels done

```

It's sometimes adds lines, but for new readers, I find it easier to get up to speed and can partition away "stuff that doesn't matter" versus getting to the crux of what the function is doing.


And then t

Reply:
Sure. At work we're using the functional approach mostly, so explicit returns are _streng verboten_ ,so I guess I default to that 😄   But that can obfuscate the code too and I see your points; it can help with readability/nesting if you don't go crazy with multiple explicit returns, agreed.

Reply:
Indeed :sweat_smile:

It's a style that DOSBox as (tended) to use (amung several); so we tend to use it just as a general rule -- but if you'd prefer not, it's not a big deal! (especially if it's in an area you'd like to overhaul more, and basically own - similar to the mouse area like @FeralChild64's has taken on.)
 
But yeah, I remember using another style, "only one return per function", it's superficially advertised as reducing branches (but instead, programmers and bury the branches and use temp variables) lol; but I suppose with time, it encourages moving branches to separate functions and making each one of them "one-shot in/out" style.

Ahh, good times.